programs/bpf_loader/src/lib.rs

```diff
@@ -145,18 +144,15 @@ pub fn process_instruction(
 
         info!("Call BPF program");
         match vm.execute_program(parameter_bytes.as_slice(), &[], &[heap_region]) {
-            Ok(status) => match u32::try_from(status) {
-                Ok(status) => {
-                    if status > 0 {
-                        warn!("BPF program failed: {}", status);
-                        return Err(InstructionError::CustomError(status));
-                    }
+            Ok(status) => {
+                // ignore upper 32bits if any, programs only return lower 32bits
```

Comment:
Could we change this to handle u64 and use the upper 32bits for custom errors to help avoid conflicting errors?

Reply:
There is a comment in the code talking about how 64-bit return values are not .js friendly.  @mvines Want to comment on this?

Reply:
I added that comment. It's just that we don't want to make JS clients handle 64bit error codes. If we split up the u64 before (top 32 for custom, bottom u32 for reserved), that's totally fine

Reply:
That might be more confusing, what if there are bits in both, is that considered a conflicting error?  Using 64 bits expands the playing field but with the same issues.  32 bits is a pretty large space for error codes and taking over the top 2 bits leaves plenty of space for user-defined errors.

Reply:
Hmm, yeah seems like `ConflictingError` ~= `BadCode`, looks unavoidable. And I think splitting u64 was a bad way to frame things.

If we are aiming for less confusion. I think that making it impossible to get conflicting error codes when using the SDK api is the way to go. Right now, it's possible because of the shared `u32` space. I'm suggesting lifting the built in error codes out of u32 and into u64.

If a dev chooses to not use our SDK then they could still hit BadCode if they return an error code that doesn't fit in u32 and doesn't map to our built in error codes. That's a case we basically already had handled here: https://github.com/solana-labs/solana/pull/7968/files#diff-1cd18121576d817cf07a6c56396a2cdeL155-L158

Reply:
That link isn't taking me to relevant code

Reply:
I agree, our SDK should protect against conflicting errors but I'm having a hard time grasping what you are proposing.  What would the bpf_loader side of things look like when interpreting the return code from the program/vm?  Or better yet, what does the 64bit return value space look like?

Reply:
Lame, was just referring to this code:

```rust
            Ok(status) => match u32::try_from(status) {
                Ok(status) => { /* .. */ }
                Err(e) => {
                    warn!("BPF VM encountered invalid status: {}", e);
                    return Err(InstructionError::GenericError);
                }
            },
```

Reply:
If status <= u32::MAX
  return as custom error
elif status is built in value (success or errors)
  return as built in value
else
  return badcode error

Reply:
Anything over u32::MAX is treated as reserved and if it doesn't map to a reserved value it would turn into a BadCode error which is opaque to the developer

Reply:
Isn't that just saying that bits 32 to 64bits are reserved vs this pr where 30th and 31st bits are reserved?

Reply:
Yeah, exactly. But this approach makes it impossible to get a badcode/conflicting error when using the sdk. And we have the u64 space, why not use it? Better than explaining to devs that they don't get the full u32 space IMO

Reply:
I feel like devs are always tempted to use upper bits as flags. So, we should allow that

Reply:
And C programs?

Reply:
Hmm, wasn't considering those. Could we add an entrypoint wrapper for those too like we have for Rust?

Reply:
Thought you were heading that direction.  Yeah I think we could, I'll work something up tomorrow

Reply:
The question becomes how does the user's C program return something that is idiomatic but allows us to convert to a larger bit space in the wrapper.